### DISCUSSION TITLE
duplicated type declaration files after packaging

### DISCUSSION THREAD

**Discussion by shellRaining** [katspaugh/wavesurfer.js#3851] [General]
```plaintext
.
├── LICENSE
├── README.md
├── dist
│   ├── base-plugin.d.ts
│   ├── base-plugin.js
│   ├── decoder.d.ts
│   ├── decoder.js
│   ├── dom.d.ts
│   ├── dom.js
│   ├── draggable.d.ts
│   ├── draggable.js
│   ├── event-emitter.d.ts
│   ├── event-emitter.js
│   ├── fetcher.d.ts
│   ├── fetcher.js
│   ├── player.d.ts
│   ├── player.js
│   ├── plugins
│   │   ├── base-plugin.d.ts
│   │   ├── decoder.d.ts
│   │   ├── dom.d.ts
│   │   ├── draggable.d.ts
│   │   ├── event-emitter.d.ts
│   │   ├── fetcher.d.ts
│   │   ├── player.d.ts
│   │   ├── plugins
│   │   │   ├── ...d.ts
│   │   │   └── zoom.d.ts
│   │   ├── ...
│   │   ├── zoom.cjs
│   │   ├── zoom.d.ts
│   │   ├── zoom.esm.js
│   │   ├── zoom.js
│   │   └── zoom.min.js
│   ├── renderer.d.ts
│   ├── renderer.js
│   ├── timer.d.ts
│   ├── timer.js
│   ├── types.d.ts
│   ├── wavesurfer.cjs
│   ├── wavesurfer.d.ts
│   ├── wavesurfer.esm.js
│   ├── wavesurfer.js
│   ├── wavesurfer.min.js
│   ├── webaudio.d.ts
│   └── webaudio.js
└── package.json
```

This is the directory structure after being bundled with Rollup. It can be observed that there are many duplicate type declaration files, such as `dist/base-plugin.d.ts`, `dist/plugins/base-plugin.d.ts`, and there exists a duplicated plugins fold in `dist/plugins`, I'm not very clear about the purpose of this packaging,  may I ask why config like this?

And when I clone this repo and run build script, there occur a error, `not found glob`, I check `package.json` and not found a package named glob, I guess there maybe miss uploading this module.

**Comment by katspaugh** [comment]
Rollup cannot build all files at once so I had to split it into two build passes: one for the main lib and one for the plugins. Plugins import some of the same things as the main lib but rollup doesn’t know about that so it makes two copies. Maybe there’s a way to somehow unify the build config and do it all in one pass.

**Reply by shellRaining** [reply]
https://github.com/shellRaining/wavesurfer.js/commit/78d7af5803051d2e17484f3e0ef9109c61a5bb61

In this commit I add config for rollup typescript plugin, when execute `npm run build`, the tsc will generate type declaration file, then rollup will compile typescript file into `esm`, `umd`... format target file, the output dir structure will be

```plaintext
.
├── base-plugin.d.ts
├── base-plugin.js
├── decoder.d.ts
├── decoder.js
├── ...
├── plugins
│   ├── ...
│   ├── zoom.cjs
│   ├── zoom.d.ts
│   ├── zoom.esm.js
│   ├── zoom.js
│   └── zoom.min.js
├── renderer.d.ts
├── renderer.js
├── timer.d.ts
├── timer.js
├── types.d.ts
├── wavesurfer.cjs
├── wavesurfer.d.ts
├── wavesurfer.esm.js
├── wavesurfer.js
├── wavesurfer.min.js
├── webaudio.d.ts
└── webaudio.js
```

Do you think this setup is okay? Is there anything I might have missed?

**Reply by katspaugh** [reply]
Looks good!

**Reply by shellRaining** [reply]
great! I'll take a pr if there are no other problem~
